Classes/EventStore/AbstractEventSourcedRepository.php

```diff
@@ -59,23 +57,19 @@ public function __construct()
 
     /**
      * @param string $identifier
-     * @return AggregateRootInterface
+     * @return EventSourcedAggregateRootInterface
      * @throws AggregateRootNotFoundException
      */
-    public function findByIdentifier($identifier)
+    final public function findByIdentifier($identifier)
     {
-        $streamName = $this->streamNameResolver->getStreamNameForAggregateTypeAndIdentifier($this->aggregateClassName, $identifier);
-        try {
-            $eventStream = $this->eventStore->get($streamName);
-        } catch (EventStreamNotFoundException $e) {
-            return null;
-        }
-
         if (!class_exists($this->aggregateClassName)) {
             throw new AggregateRootNotFoundException(sprintf("Could not reconstitute the aggregate root %s because its class '%s' does not exist.", $identifier, $this->aggregateClassName), 1474454928115);
         }
 
-        $aggregateRoot = unserialize('O:' . strlen($this->aggregateClassName) . ':"' . $this->aggregateClassName . '":1:{s:13:"' . chr(0) . '*' . chr(0) . 'identifier";s:36:"' . $identifier . '";}');
+        $streamName = $this->streamNameResolver->getStreamNameForAggregateTypeAndIdentifier($this->aggregateClassName, $identifier);
+        $eventStream = $this->eventStore->get($streamName);
+
+        $aggregateRoot = unserialize('O:' . strlen($this->aggregateClassName) . ':"' . $this->aggregateClassName . '":2:{s:13:"' . chr(0) . '*' . chr(0) . 'identifier";s:' . strlen($identifier) . ':"' . $identifier . '";s:10:"' . chr(0) . '*' . chr(0) . 'version";i:' . (string)$eventStream->getVersion() . ';}');
```

Comment:
I don't exactly get your problem with the current implementation.
Granted, the `unserialize()` trick is a bit "funky", but we use the very same mechanism to reconstitute entities from persistence since years.

Anyways, I agree that we can probably improve that part.. Let's discuss it asap, I need to let it sink in first